lib/improver/nbhood_percentiles.py

```diff
@@ -0,0 +1,375 @@
+        """
+        self.percentiles = percentiles
+
+    def __repr__(self):
+        """Represent the configured class instance as a string."""
+        result = ('<CircularKernelNumpy: percentiles: {}>')
+        return result.format(self.percentiles)
+
+    def run(self, cube, ranges):
+        """
+        Method to apply a circular kernel to the data within the input cube in
+        order to derive percentiles over the kernel.
+
+        Parameters
+        ----------
+        cube : Iris.cube.Cube
+            Cube containing array to apply processing to.
+        ranges : Int
+            Number of grid cells in the x and y direction used to create
+            the kernel.
+
+        Returns
+        -------
+        outcube : Iris.cube.Cube
+            Cube containing the percentile fields.
+            Has percentile as an added dimension.
+
+        """
+        # Take data array and identify X and Y axes indices
+        try:
+            for coord_name in ['projection_x_coordinate',
+                               'projection_y_coordinate']:
+                cube.coord(coord_name)
+        except CoordinateNotFoundError:
+            raise ValueError("Invalid grid: projection_x/y coords required")
+        ranges = int(ranges)
+        if ranges < 1:
+            raise ValueError("Range size too small. {} < 1".format(ranges))
+        ranges_xy = np.array([ranges]*2)
+        # Define the size of the kernel based on the number of grid cells
+        # contained within the desired radius.
+        kernel = np.ones([int(1 + x * 2) for x in ranges_xy])
+        # Create an open multi-dimensional meshgrid.
+        open_grid = np.array(np.ogrid[tuple([slice(-x, x+1) for x in ranges_xy])])
+        # Always generate kernel in unweighted mode as later logic doesn't make sense otherwise
+        mask = np.reshape(
+            np.sum(open_grid**2) > np.prod(ranges_xy), np.shape(kernel))
+        kernel[mask] = 0.
+
+        # Loop over each 2D slice to reduce memory demand and derive percentiles
+        # on the kernel. Will return an extra dimension.
+        pctcubelist = iris.cube.CubeList()
+        for slice_2d in cube.slices(['projection_x_coordinate',
+                                     'projection_y_coordinate']):
+            # Create a 1D data array padded with repeats of the local boundary mean.
+            padded = np.pad(slice_2d.data, ranges, mode='mean', stat_length=ranges)
+            padshape = np.shape(padded) # Store size to make unflatten easier
+            padded = padded.flatten()
+            # Add 2nd dimension with each point's neighbourhood points along it
+            nbhood_slices = [np.roll(padded, padshape[1]*j+i) for i in range(-ranges, ranges+1) for j in range(-ranges, ranges+1) if kernel[...,i+ranges, j+ranges]>0.]
+            # Collapse this dimension into percentiles (a new 2nd dimension)
+            perc_data = np.percentile(nbhood_slices, self.percentiles, axis=0)
+            # Return to 3D
```

Comment:
I think the percentile plugin will be using the np one eventually, so hopefully this will be ok.